File: exts/jphp-gui-game-ext/src/main/java/org/develnext/jphp/ext/game/classes/UXGameEntity.java
package org.develnext.jphp.ext.game.classes;

import javafx.event.EventHandler;
import javafx.scene.Node;
import org.develnext.jphp.ext.game.GameExtension;
import org.develnext.jphp.ext.game.support.CollisionEvent;
import org.develnext.jphp.ext.game.support.GameEntity2D;
import org.develnext.jphp.ext.game.support.Vec2d;
import php.runtime.annotation.Reflection.*;
import php.runtime.env.Environment;
import php.runtime.invoke.Invoker;
import php.runtime.lang.BaseWrapper;
import php.runtime.reflection.ClassEntity;

@Namespace(GameExtension.NS)
public class UXGameEntity extends BaseWrapper<GameEntity2D> {
    interface WrappedInterface {
        @Property Node node();
        @Property GameEntity2D.BodyType bodyType();
        @Property @Nullable Vec2d gravity();

        @Property GameEntity2D.SolidType solidType();
        @Property boolean solid();
        @Property boolean platform();
        @Property boolean active();

        @Property double gravityX();
        @Property double gravityY();

        @Property double velocityX();
        @Property double velocityY();

        @Property Vec2d velocity();

        @Property Vec2d angleSpeed();
        @Property double hspeed();
        @Property double vspeed();
        @Property double speed();
        @Property double direction();

        void setPolygonFixture(Vec2d[] points);
        void setRectangleFixture(double width, double height);
        void setEllipseFixture(double width, double height);
        void setCircleFixture(double radius);

        boolean isPlaceFree(double x, double y);
    }

    public UXGameEntity(Environment env, GameEntity2D wrappedObject) {
        super(env, wrappedObject);
    }

    public UXGameEntity(Environment env, ClassEntity clazz) {
        super(env, clazz);
    }

    @Signature
    public void __construct(String entityType, Node node) {
        __wrappedObject = new GameEntity2D(entityType, node);
    }

    @Getter
    public UXGameScene getGameScene(Environment env) {
        return getWrappedObject().getScene() == null ? null : new UXGameScene(env, getWrappedObject().getScene());
    }

    @Signature
    public void setCollisionHandler(String entityType, @Nullable final Invoker handler) {
        if (handler == null) {
            getWrappedObject().setCollisionHandler(entityType, null);
        } else {
            getWrappedObject().setCollisionHandler(entityType, new EventHandler<CollisionEvent>() {
                @Override
                public void handle(CollisionEvent event) {
                    handler.callAny(event);
                }
            });
        }
    }
}

File: exts/jphp-gui-game-ext/src/main/resources/JPHP-INF/sdk/php/game/UXGameEntity.php
<?php
namespace php\game;

use php\gui\UXNode;
use php\gui\UXParent;

/**
 * Class UXGameObject
 * @package php\game
 * @packages game, javafx
 */
class UXGameEntity
{
    /**
     * @var UXNode
     */
    public $node;

    /**
     * @var bool
     */
    public $active = true;

    /**
     * @var float
     */
    public $x;

    /**
     * @var float
     */
    public $y;

    /**
     * @var float
     */
    public $centerX;

    /**
     * @var float
     */
    public $centerY;

    /**
     * @var string STATIC, DYNAMIC, KINEMATIC
     */
    public $bodyType = 'STATIC';

    /**
     * If null, using scene gravity
     * @var array|null [x, y]
     */
    public $gravity = null;

    /**
     * @var float
     */
    public $gravityX = 0.0;

    /**
     * @var float
     */
    public $gravityY = 0.0;

    /**
     * @readonly
     * @var UXGameScene
     */
    public $gameScene = null;

    /**
     * @var array
     */
    public $velocity = [0, 0];

    /**
     * @var float
     */
    public $velocityX = 0.0;

    /**
     * @var float
     */
    public $velocityY = 0.0;

    /**
     * @var array alias of velocity property
     */
    public $angleSpeed = [0, 0];

    /**
     * @var float angle speed
     */
    public $speed = 0.0;

    /**
     * @var float angle for speed, from 0 to 360
     */
    public $direction = 0.0;

    /**
     * @var float alias of velocityX
     */
    public $hspeed = 0.0;

    /**
     * @var float alias of velocityY
     */
    public $vspeed = 0.0;

    /**
     * @var string NONE, PLATFORM, PHYSIC
     */
    public $solidType = 'NONE';

    /**
     * @readonly
     * @var bool
     */
    public $solid = false;

    /**
     * @readonly
     * @var bool
     */
    public $platform = false;

    /**
     * @param string $entityType
     * @param UXNode $node
     */
    public function __construct($entityType, UXNode $node)
    {
    }

    /**
     * @param array $points  [[x,y], [x,y], [x,y], [x,y] ...]
     */
    public function setPolygonFixture(array $points)
    {
    }

    /**
     * @param float $width
     * @param float $height
     */
    public function setRectangleFixture($width, $height)
    {
    }

    /**
     * @param float $width
     * @param float $height
     */
    public function setEllipseFixture($width, $height)
    {
    }

    /**
     * @param float $radius
     */
    public function setCircleFixture($radius)
    {
    }

    /**
     * @param string $entityType
     * @param callable|null $handler
     */
    public function setCollisionHandler($entityType, callable $handler)
    {
    }
}